File: cpp/data_structures/two_chapter/sqrt.c
#include <iostream>

using namespace std;

double sqrt(const double &d, const double precision) {
	double k = d;
	while (1) {
		if (k * k > d - precision && k * k < d + precision)
			break;
		k = 0.5 * (k + d / k);	//newton algorithm
	}
	return k;
}

int main(int argc, char** argv)
{
	double d;
	while (cout << "please input a number:", cin >> d)
		cout << endl << "the square root is: " << sqrt(d, 0.1) << endl;

	return 0;
}

File: go/data_structure/1_chapter/sqrt.go
package main

import (
	"fmt"
	"math"
)

func main() {
	var n float64
	var err error = nil
	for err == nil {
		fmt.Print("please input a number:")
		_, err = fmt.Scan(&n)
		fmt.Println("the number( " + fmt.Sprintf("%f", n) + " )square root is : " + fmt.Sprintf("%f", sqrt(n)))
	}
}

// return the result of square root
// newton algorithm
func sqrt(n float64) float64 {
	res := n
	for math.Abs(res*res-n) > 0.1 {
		res = (n/res + res) / 2
	}
	return res
}
